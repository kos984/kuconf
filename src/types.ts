// import ValidatorJs from '../../ValidatorJs';
export type StringsValue<T> = {
  [P in keyof T]: StringsValue<T[P]> | string;
};
export interface ILogger {
  log(...args: any[]): any;
  warn(...args: any[]): any;
}
export interface IConfigOptions {
  allowGet: boolean;
  getSeparator: string;
  logger: ILogger;
}
export interface IParseEnvParams {
  /** empty by default */
  prefix?: string;
  /** __ by default */
  delimiter?: string;
  /** if true will replace all '_' to '', false by default */
  ignoreOneLodash?: boolean;
  doNotWarnIfKeyOverridden?: boolean;
}

export interface IGenerateEnvParams {
  /** empty by default */
  prefix?: string;
  /** __ by default */
  delimiter?: string;
}

export interface IOmitNotValidatedProps {
  /** true by default */
  logOmitted: boolean;
}
