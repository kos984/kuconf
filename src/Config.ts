import * as _ from 'lodash';
import ProxyHandler, { validate as partialValidate } from './ProxyHandler';
import { IConfigOptions, ILogger } from './types';
import Validator from './Validator';

const privateStoreKey = Symbol('config');

export default class Config<ConfigSchema> {

  /** ValidatorJs rules */
  // public schema: any = {};

  protected config: object = {};
  protected configObject: ConfigSchema;

  protected handler: ProxyHandler;
  protected logger: ILogger;
  protected validation: Validator.Validator<any>;
  // protected parsedKeyPaths: string[] = [];
  protected options: IConfigOptions;

  protected [privateStoreKey]: Map<any, any>;

  constructor(config: object, options?: Partial<IConfigOptions>) {
    this.options = {
      get: {
        allowed: false,
        separator: '.',
      },
      logger: console,
      validation: null,
      ...options};
    this.logger = this.options.logger;
    this.config = this.options.caseSensitive ? config : this.objectToLower(config);
    this.handler = new ProxyHandler(this, { caseSensitive: this.options.caseSensitive });
  }

  public merge(obj: Partial<ConfigSchema>): Config<ConfigSchema> {
    this.config = _.merge(this.config, this.options.caseSensitive ? obj : this.objectToLower(obj));
    // reset validation;
    this.validation = null;
    this.initValidator();
    return this;
  }

  public get(path: string, defaultValue?: any) {
    if (!this.options.get.allowed) {
      throw new Error('get is not allowed');
    }
    const config = this.getConfig();
    path = path.replace(new RegExp(this.options.get.separator, 'g'), '.');
    return _.get(config, path, defaultValue);
  }

  public getConfig(): ConfigSchema {
    if (this.configObject) {
      return this.configObject;
    }
    this.initValidator();
    this.configObject = new Proxy(this.config as any, this.handler);
    return this.configObject;
  }

  public validate(confPart?: any) {
    this.initValidator();
    if (arguments.length > 0 && (!confPart || !confPart[partialValidate])) {
      throw new Error('only objects allowed for validation');
    }
    const errors = arguments.length !== 0 ? confPart[partialValidate]() : this.validation.errors;
    if (errors && Object.keys(errors.errors).length) {
      throw new Error(JSON.stringify(errors));
    }
    return this;
  }

  public getValidationErrors(path?: string) {
    this.initValidator();
    if (!this.validation.errors) {
      return null;
    }
    if (!path) {
      return this.validation.errors;
    } else if (!this.options.caseSensitive) {
      path = path.toLowerCase();
    }
    const result: any = {
      errors: {},
    };
    let found = false;
    Object.keys(this.validation.errors.errors).forEach(key => {
      if (key.indexOf(path) === 0) {
        found = true;
        result.errors[key] = (this.validation.errors.errors as any)[key];
      }
    });
    return found ? result : null;
  }

  protected initValidator() {
    if (!this.options.validation || this.validation) {
      return;
    }
    this.validation = new Validator(this.config, this.prepareRules(this.options.validation.rules));
    Object.keys(this.validation.rules).forEach( key => {
      const newKey = this.options.caseSensitive ? key : key.toLowerCase();
      if (key !== newKey) {
        this.validation.rules[newKey] = this.validation.rules[key];
        delete this.validation.rules[key];
      }
    });
    this.validation.fails();
  }

  // in progress, so for not not public
  /*
  private omitNotValidatedProps(params: IOmitNotValidatedProps) {
    params = { logOmitted: true, ...params };
    this.initValidator();
    const newConfig = {};
    const rules: { [key: string]: boolean} = {};
    Object.keys(this.validation.rules).forEach(rule => (rules[rule.toLowerCase()] = true));
    for (const path of this.parsedKeyPaths) {
      if (!rules[path]) {
        if (params.logOmitted) {
          this.logger.warn(`path: ${path} is not under validation and will be omitted`);
        }
      } else {
        _.set(newConfig, path, _.get(this.config, path));
      }
    }
    // add default keys
    Object.keys(rules).forEach(path => {
      if (_.has(this.config, path)) {
        _.set(newConfig, path, _.get(this.config, path));
      }
    });

    this.config = newConfig as ConfigSchema;

    return this;
  }

  */

  protected objectToLower(obj: any, result: any = {}): any {
    if (Array.isArray(obj)) {
      return obj.map(o => this.objectToLower(o));
    } else if (typeof obj === 'object') {
      Object.keys(obj).forEach((key: string) => {
        const path = typeof key === 'string' ? key.toLowerCase() : key;
        const value = this.objectToLower(obj[key], result[key] ? result[key] : {});
        result[path] = value;
      });
      return result;
    }
    return obj;
  }

  protected prepareRules(schema: any) {
    const result: { [key: string]: string } = {};
    const parse = (obj: any, start?: string) => {
      Object.keys(obj).forEach(key => {
        const fullKey = start ? start + '.' : '';
        const value = obj[key];
        if (typeof value === 'string') {
          result[`${fullKey}${key}`] = value;
        } else if (Array.isArray(value)) {
          parse(value[0], `${fullKey}${key}.*`);
        } else {
          parse(value, `${fullKey}${key}`);
        }
      });
    };
    parse(schema);
    return result;
  }

}
