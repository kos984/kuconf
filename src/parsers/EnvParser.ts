import * as _ from 'lodash';

export interface IEnvParserParams {
  /** @default '' */
  prefix: string;
  /** @default '__' */
  delimiter: string;
  /**
   * if true will replace all '_' to ''
   * works only if format function not passed
   * @default false
   */
  ignoreOneLodash: boolean;
  /**
   * works only if format function not passed
   * @default true
   */
  lower: boolean;
  /**
   * format key, buy default make key lower and delete all '_' symbols
   */
  format: (key: string) => string;
}

export default class EnvParser {
  protected params: IEnvParserParams;
  protected parsed: any = {};
  public constructor(params?: Partial<IEnvParserParams>) {
    this.params = {
      delimiter: '__',
      format: key => {
        return this.params.ignoreOneLodash ? key.replace(/_/g, '') : key.toLowerCase();
      },
      ignoreOneLodash: false,
      lower: true,
      prefix: '',
      ...params,
    };
    this.parse();
  }
  public get() {
    return this.parsed;
  }

  protected parse() {
    const { prefix, delimiter, format } = this.params;
    const parsed: any = this.parsed;
    Object.keys(process.env)
      .filter(key => (key.indexOf(prefix) === 0))
      .map(key => {
        const keyPath = key
          .replace(prefix, '')
          .split(delimiter)
          .map(subKey => format(subKey))
          .join('.');
        _.set(parsed, keyPath, process.env[key]);
      });
    return this;
  }

  /*
  public generateEnv(params?: IGenerateEnvParams): string[] {
    const { prefix, delimiter } = {
      delimiter: '__',
      prefix: '',
      ...params };
    this.initValidator();
    const rules = this.validation.rules;
    return Object.keys(rules).map((rule: string) => {
      return prefix + rule.replace(/\./g, delimiter).toUpperCase() + '=';
    });
  }
  */
}
