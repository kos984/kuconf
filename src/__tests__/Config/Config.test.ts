import './env';

import * as _ from 'lodash';
import Config from '../../Config';
import EnvParser from '../../parsers/EnvParser';
import { IConfig } from './types';
import rules from './validationRules';

const prefix = 'TEST__';

describe('Config [default settings]', () => {

  describe('toObject test', () => {
    it('toObject', () => {
      const envParser = new EnvParser({ prefix });
      const conf = new Config<IConfig>(envParser.get(), {
        caseSensitive: false,
      });
      const config = conf.getConfig();
      expect(conf.toObject(config.db)).toMatchSnapshot();
    });
  });

  describe('[options] caseSensitive', () => {
    it('config should be case-insensitive', () => {
      const envParser = new EnvParser({ prefix });
      const conf = new Config<IConfig>(envParser.get(), {
        validation: { rules },
      });
      const config = conf.getConfig();
      const c: any = config;
      expect(config.db === c.DB).toBeTruthy();
      expect(config.db.username === c.db.UserName).toBeTruthy();
      expect(config.db.database === c.db.dataBase).toBeTruthy();
      // expect(config.queue).toMatchSnapshot();
    });
    it('config should be case-sensitive', () => {
      const envParser = new EnvParser({ prefix, lower: true });
      const conf = new Config<IConfig>(envParser.get(), {
        caseSensitive: true,
        validation: { rules },
      });
      const config = conf.getConfig();
      expect(config.db).toMatchSnapshot();
      expect((config as any).DB).not.toBeDefined();
    });
  });
  describe('[options] validation', () => {
    const envParser = new EnvParser({ prefix });
    const conf = new Config<IConfig>(envParser.get(), {
      validation: { rules },
    });
    const config = conf.getConfig();
    it ('rules should be parsed', () => {
      expect((conf as any).validation.rules).toMatchSnapshot();
    });
    it ('config should be correct', () => {
      expect(config).toMatchSnapshot();
    });
    it ('root validations should not pass', () => {
      expect(() => conf.validate()).toThrowErrorMatchingSnapshot();
    });
    it('db validation should not throw errors', () => {
      expect(() => conf.validate(config.db)).not.toThrowError();
    });
    it('validation redis should throw error', () => {
      expect(() => conf.validate(config.redis.host)).toThrowErrorMatchingSnapshot();
    });
    it('validation case isensitive should return errors', () => {
      expect(conf.getValidationErrors('features.featureName')).toMatchSnapshot();
    });
    it('validation redis.host should throw error', () => {
      expect(() => conf.validate(config.redis.host)).toThrowErrorMatchingSnapshot();
    });
    it('validation redis.sentinels[0] should throw error', () => {
      expect(() => conf.validate(config.redis.sentinels[0])).toThrowErrorMatchingSnapshot();
    });
    it('validation redis.sentinels[1] should pass', () => {
      expect(() => conf.validate(config.redis.sentinels[1])).not.toThrowError();
    });
  });

  describe('[options] get', () => {
    const envParser = new EnvParser({ prefix, lower: true });
    describe('true', () => {
      const conf = new Config<IConfig>(envParser.get(), {
        get: { allowed: true, separator: ':' },
        validation: { rules },
      });
      const config = conf.getConfig();
      expect(config.db).toMatchSnapshot();
      it('conf.get should works', () => {
        expect(config.db.replication.write === conf.get('db:replication:write')).toBeTruthy();
        expect(config.db.replication.write === conf.get('db.replication.write')).toBeTruthy();
      });
      it('lodash get should works', () => {
        expect(config.db.replication.write === _.get(config, 'db.replication.write'));
      });
    });
    describe('false', () => {
      const conf = new Config<IConfig>(envParser.get(), {
        validation: { rules },
      });
      const config = conf.getConfig();
      expect(config.db).toMatchSnapshot();
      it('conf.get should not allow get', () => {
        expect(() => conf.get('db:replication:write')).toThrowErrorMatchingSnapshot();
      });
      it('lodash get should works', () => {
        expect(config.db.replication.write === _.get(config, 'db.replication.write'));
      });
    });
  });

  describe('merge', () => {
    const envParser = new EnvParser({ prefix });
    const conf = new Config<IConfig>(envParser.get(), {
      validation: { rules },
    });
    conf.merge({ db: { port: 777 } as any});
    const config = conf.getConfig();
    expect(config).toMatchSnapshot();
    conf.merge({ db: { port: '7777' } as any});
    expect(config).toMatchSnapshot();
  });
});
