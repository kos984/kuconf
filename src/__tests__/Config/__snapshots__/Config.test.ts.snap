// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
Object {
  "database": "test",
  "host": "localhost",
  "password": "password",
  "port": 3306,
  "replication": Object {
    "read": Array [
      Object {
        "host": "read0.example.com",
        "password": "password_read_0",
        "username": "username_read_0",
      },
      Object {
        "host": "read1.example.com",
        "password": "password_read_1",
        "username": "username_read_1",
      },
    ],
    "write": Object {
      "host": "write.example.com",
      "password": "password_write",
      "username": "username_write",
    },
  },
  "username": "username",
}
`;

exports[` 2`] = `
Object {
  "database": "test",
  "host": "localhost",
  "password": "password",
  "port": 3306,
  "replication": Object {
    "read": Array [
      Object {
        "host": "read0.example.com",
        "password": "password_read_0",
        "username": "username_read_0",
      },
      Object {
        "host": "read1.example.com",
        "password": "password_read_1",
        "username": "username_read_1",
      },
    ],
    "write": Object {
      "host": "write.example.com",
      "password": "password_write",
      "username": "username_write",
    },
  },
  "username": "username",
}
`;

exports[` 3`] = `
Object {
  "db": Object {
    "database": "test",
    "host": "localhost",
    "password": "password",
    "port": 777,
    "replication": Object {
      "read": Array [
        Object {
          "host": "read0.example.com",
          "password": "password_read_0",
          "username": "username_read_0",
        },
        Object {
          "host": "read1.example.com",
          "password": "password_read_1",
          "username": "username_read_1",
        },
      ],
      "write": Object {
        "host": "write.example.com",
        "password": "password_write",
        "username": "username_write",
      },
    },
    "username": "username",
  },
  "features": Object {
    "featurename": Object {
      "enabled": false,
    },
  },
  "redis": Object {
    "sentinels": Array [
      Object {
        "host": "localhost",
        "port": "26379",
        "test": false,
      },
      Object {
        "host": "localhost",
        "port": "26379",
        "test": true,
      },
    ],
  },
}
`;

exports[` 4`] = `
Object {
  "db": Object {
    "database": "test",
    "host": "localhost",
    "password": "password",
    "port": 7777,
    "replication": Object {
      "read": Array [
        Object {
          "host": "read0.example.com",
          "password": "password_read_0",
          "username": "username_read_0",
        },
        Object {
          "host": "read1.example.com",
          "password": "password_read_1",
          "username": "username_read_1",
        },
      ],
      "write": Object {
        "host": "write.example.com",
        "password": "password_write",
        "username": "username_write",
      },
    },
    "username": "username",
  },
  "features": Object {
    "featurename": Object {
      "enabled": false,
    },
  },
  "redis": Object {
    "sentinels": Array [
      Object {
        "host": "localhost",
        "port": "26379",
        "test": false,
      },
      Object {
        "host": "localhost",
        "port": "26379",
        "test": true,
      },
    ],
  },
}
`;

exports[`Config [default settings] [options] caseSensitive config should be case-sensitive 1`] = `
Object {
  "database": "test",
  "host": "localhost",
  "password": "password",
  "port": 3306,
  "replication": Object {
    "read": Array [
      Object {
        "host": "read0.example.com",
        "password": "password_read_0",
        "username": "username_read_0",
      },
      Object {
        "host": "read1.example.com",
        "password": "password_read_1",
        "username": "username_read_1",
      },
    ],
    "write": Object {
      "host": "write.example.com",
      "password": "password_write",
      "username": "username_write",
    },
  },
  "username": "username",
}
`;

exports[`Config [default settings] [options] get false conf.get should not allow get 1`] = `"get is not allowed"`;

exports[`Config [default settings] [options] validation config should be correct 1`] = `
Object {
  "db": Object {
    "database": "test",
    "host": "localhost",
    "password": "password",
    "port": 3306,
    "replication": Object {
      "read": Array [
        Object {
          "host": "read0.example.com",
          "password": "password_read_0",
          "username": "username_read_0",
        },
        Object {
          "host": "read1.example.com",
          "password": "password_read_1",
          "username": "username_read_1",
        },
      ],
      "write": Object {
        "host": "write.example.com",
        "password": "password_write",
        "username": "username_write",
      },
    },
    "username": "username",
  },
  "features": Object {
    "featurename": Object {
      "enabled": false,
    },
  },
  "redis": Object {
    "sentinels": Array [
      Object {
        "host": "localhost",
        "port": "26379",
        "test": false,
      },
      Object {
        "host": "localhost",
        "port": "26379",
        "test": true,
      },
    ],
  },
}
`;

exports[`Config [default settings] [options] validation root validations should not pass 1`] = `"{\\"errors\\":{\\"redis.sentinels.0.test\\":[\\"The redis.sentinels.0.test must be a true or false\\"],\\"features.featurename.key\\":[\\"The features.featurename.key field is required.\\"]}}"`;

exports[`Config [default settings] [options] validation rules should be parsed 1`] = `
Object {
  "db.database": Array [
    Object {
      "name": "string",
    },
  ],
  "db.host": Array [
    Object {
      "name": "required_with",
      "value": "db",
    },
    Object {
      "name": "string",
    },
  ],
  "db.password": Array [
    Object {
      "name": "string",
    },
  ],
  "db.port": Array [
    Object {
      "name": "integer",
    },
    Object {
      "name": "cast",
      "value": "number",
    },
  ],
  "db.replication.read.0.host": Array [
    Object {
      "name": "string",
    },
  ],
  "db.replication.read.0.password": Array [
    Object {
      "name": "string",
    },
  ],
  "db.replication.read.0.port": Array [
    Object {
      "name": "integer",
    },
    Object {
      "name": "cast",
      "value": "number",
    },
  ],
  "db.replication.read.0.username": Array [
    Object {
      "name": "string",
    },
  ],
  "db.replication.read.1.host": Array [
    Object {
      "name": "string",
    },
  ],
  "db.replication.read.1.password": Array [
    Object {
      "name": "string",
    },
  ],
  "db.replication.read.1.port": Array [
    Object {
      "name": "integer",
    },
    Object {
      "name": "cast",
      "value": "number",
    },
  ],
  "db.replication.read.1.username": Array [
    Object {
      "name": "string",
    },
  ],
  "db.replication.write.host": Array [
    Object {
      "name": "string",
    },
  ],
  "db.replication.write.password": Array [
    Object {
      "name": "string",
    },
  ],
  "db.replication.write.port": Array [
    Object {
      "name": "number",
    },
    Object {
      "name": "cast",
      "value": "number",
    },
  ],
  "db.replication.write.username": Array [
    Object {
      "name": "string",
    },
  ],
  "db.username": Array [
    Object {
      "name": "string",
    },
  ],
  "features.featurename.enabled": Array [
    Object {
      "name": "bool",
    },
    Object {
      "name": "cast",
      "value": "boolean,false",
    },
  ],
  "features.featurename.key": Array [
    Object {
      "name": "required",
    },
    Object {
      "name": "string",
    },
  ],
  "queue.first.accesskeyid": Array [
    Object {
      "name": "string",
    },
  ],
  "queue.first.endpoint": Array [
    Object {
      "name": "string",
    },
  ],
  "queue.first.queuename": Array [
    Object {
      "name": "string",
    },
  ],
  "queue.first.queueurl": Array [
    Object {
      "name": "string",
    },
  ],
  "queue.first.region": Array [
    Object {
      "name": "string",
    },
  ],
  "queue.first.secretaccesskey": Array [
    Object {
      "name": "string",
    },
  ],
  "queue.second.accesskeyid": Array [
    Object {
      "name": "string",
    },
  ],
  "queue.second.endpoint": Array [
    Object {
      "name": "string",
    },
  ],
  "queue.second.queuename": Array [
    Object {
      "name": "string",
    },
  ],
  "queue.second.queueurl": Array [
    Object {
      "name": "string",
    },
  ],
  "queue.second.region": Array [
    Object {
      "name": "string",
    },
  ],
  "queue.second.secretaccesskey": Array [
    Object {
      "name": "string",
    },
  ],
  "redis.host": Array [
    Object {
      "name": "string",
    },
  ],
  "redis.port": Array [
    Object {
      "name": "integer",
    },
  ],
  "redis.sentinels.0.host": Array [
    Object {
      "name": "string",
    },
  ],
  "redis.sentinels.0.port": Array [
    Object {
      "name": "integer",
    },
  ],
  "redis.sentinels.0.test": Array [
    Object {
      "name": "required",
    },
    Object {
      "name": "bool",
    },
    Object {
      "name": "cast",
      "value": "boolean,false",
    },
  ],
  "redis.sentinels.1.host": Array [
    Object {
      "name": "string",
    },
  ],
  "redis.sentinels.1.port": Array [
    Object {
      "name": "integer",
    },
  ],
  "redis.sentinels.1.test": Array [
    Object {
      "name": "required",
    },
    Object {
      "name": "bool",
    },
    Object {
      "name": "cast",
      "value": "boolean,false",
    },
  ],
}
`;

exports[`Config [default settings] [options] validation validation case isensitive should return errors 1`] = `
Object {
  "errors": Object {
    "features.featurename.key": Array [
      "The features.featurename.key field is required.",
    ],
  },
}
`;

exports[`Config [default settings] [options] validation validation redis should throw error 1`] = `"only objects allowed for validation"`;

exports[`Config [default settings] [options] validation validation redis.host should throw error 1`] = `"only objects allowed for validation"`;

exports[`Config [default settings] [options] validation validation redis.sentinels[0] should throw error 1`] = `"{\\"errors\\":{\\"redis.sentinels.0.test\\":[\\"The redis.sentinels.0.test must be a true or false\\"]}}"`;
